nullaway/src/main/java/com/uber/nullaway/fixserialization/Serializer.java

```diff
@@ -0,0 +1,119 @@
+import com.uber.nullaway.fixserialization.out.ErrorInfo;
+import com.uber.nullaway.fixserialization.out.SuggestedFixInfo;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+/**
+ * Serializer class where all generated files in Fix Serialization package is created through APIs
+ * of this class.
+ */
+public class Serializer {
+  /** Path to write errors. */
+  public final Path ERROR;
+  /** Path to write suggested fix metadata. */
+  public final Path SUGGEST_FIX;
+
+  public Serializer(FixSerializationConfig config) {
+    String outputDirectory = config.outputDirectory;
+    this.ERROR = Paths.get(outputDirectory, "errors.tsv");
+    this.SUGGEST_FIX = Paths.get(outputDirectory, "fixes.tsv");
+    reset(config);
+  }
+
+  /**
+   * Appends the string representation of the {@link SuggestedFixInfo}.
+   *
+   * @param suggestedFixInfo SuggestedFixInfo object.
+   * @param enclosing Flag to control if enclosing method and class should be included.
+   */
+  public void serializeSuggestedFixInfo(SuggestedFixInfo suggestedFixInfo, boolean enclosing) {
+    if (enclosing) {
+      suggestedFixInfo.findEnclosing();
+    }
+    appendToFile(suggestedFixInfo.tabSeparatedToString(), SUGGEST_FIX);
+  }
+
+  /**
+   * Appends the string representation of the {@link ErrorMessage}.
+   *
+   * @param errorInfo ErrorMessage object.
+   */
+  public void serializeErrorInfo(ErrorInfo errorInfo) {
+    errorInfo.findEnclosing();
+    appendToFile(errorInfo.tabSeparatedToString(), ERROR);
+  }
+
+  /** Cleared the content of the file if exists and writes the header in the first line. */
+  private void resetFile(Path path, String header) {
+    try {
+      Files.deleteIfExists(path);
+      OutputStream os = new FileOutputStream(path.toFile());
+      header += "\n";
+      os.write(header.getBytes(Charset.defaultCharset()), 0, header.length());
+      os.flush();
+      os.close();
+    } catch (IOException e) {
+      throw new RuntimeException(
+          "Could not finish resetting File at Path: " + path + ", Exception: " + e);
+    }
+  }
+
+  /** Resets every file which will be re-generated in the new run of NullAway. */
+  private void reset(FixSerializationConfig config) {
+    try {
+      Files.createDirectories(Paths.get(config.outputDirectory));
+      if (config.suggestEnabled) {
+        resetFile(SUGGEST_FIX, SuggestedFixInfo.header());
+      }
+      resetFile(ERROR, ErrorInfo.header());
+    } catch (IOException e) {
+      throw new RuntimeException("Could not finish resetting serializer: " + e);
+    }
+  }
+
+  private void appendToFile(String row, Path path) {
+    OutputStream os;
+    if (row == null || row.equals("")) {
+      return;
+    }
+    row = row.replaceAll("\\R+", " ") + "\n";
+    try {
+      os = new FileOutputStream(path.toFile(), true);
```

Comment:
It's a bit awkward and inefficient to be opening and closing a stream every time you want to write a line to the file.  I guess you did it this way because there is no way to call some kind of "finalize" method to know when to flush and close a single stream?

In any case, this kind of code should be written using try-with-resources blocks to avoid leaks: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

Reply:
Hi @msridhar, Yes exactly, that is the reason why I had to write it this way. Thank you for the suggestion, applied it here: a05aa6a2d4f4e64da19f0c03e38967d1bd37ec0e